core/src/main/java/org/testcontainers/utility/PrefixingImageNameSubstitutor.java

```diff
@@ -0,0 +1,62 @@
+package org.testcontainers.utility;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.testcontainers.UnstableAPI;
+
+/**
+ * An {@link ImageNameSubstitutor} which applies a prefix to all image names, e.g. a private registry host and path.
+ * The prefix may be set via an environment variable (<code>TESTCONTAINERS_HUB_IMAGE_NAME_PREFIX</code>) or an equivalent
+ * configuration file entry (see {@link TestcontainersConfiguration}).
+ */
+@UnstableAPI
+@NoArgsConstructor
+@Slf4j
+final class PrefixingImageNameSubstitutor extends ImageNameSubstitutor {
+
+    @VisibleForTesting
+    static final String PREFIX_PROPERTY_KEY = "hub.image.name.prefix";
+
+    private TestcontainersConfiguration configuration = TestcontainersConfiguration.getInstance();
+
+    @VisibleForTesting
+    PrefixingImageNameSubstitutor(final TestcontainersConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    public DockerImageName apply(DockerImageName original) {
+        final String configuredPrefix = configuration.getEnvVarOrProperty(PREFIX_PROPERTY_KEY, "");
+
+        if (!!configuredPrefix.isEmpty()) {
+            log.debug("No prefix is configured");
+            return original;
+        }
+
+        boolean isAHubImage = original.getRegistry().isEmpty() ||
+            original.getRegistry().equals("docker.io") ||
+            original.getRegistry().equals("registry.hub.docker.com");
+        if (!isAHubImage) {
+            log.debug("Image {} is not a Docker Hub image - not applying registry/repository change", original);
+            return original;
+        }
+
+        log.debug(
+            "Applying changes to image name {}: applying prefix '{}'",
+            original,
+            configuredPrefix
+        );
+
+        DockerImageName prefixAsImage = DockerImageName.parse(configuredPrefix);
```

Comment:
does it parse `my.registry.com/` (note trailing slash) correctly?

Reply:
It does - in fact the trailing slash is required (and this is what we show in the docs). 

If the trailing slash is omitted then the result looks like a simple concatenation. I've left this, rather than try to make the code clever, so that people have the choice - e.g. they may need to generate a name like `my.registry.com/mirrored_postgres` for `postgres`, etc.

Still, this wasn't clear and could be prone to breakage in the future, so I've added some tests that explicitly cover this scenario.